// u-wave-nodejs-client/lib/modules/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
class Auth {
    constructor(uw, onAuthenticated) {
        this.uw = uw;
        this.onAuthenticated = onAuthenticated;
    }
    getCurrentUser() {
        return this.uw
            .get('/auth')
            .then((response) => {
            if (!response.data)
                return null;
            return helpers_1.parseDates(response.data, Auth.USER_DATE_FIELDS);
        });
    }
    getSocketToken() {
        return this.uw
            .get('/auth/socket')
            .then((res) => res.data.socketToken);
    }
    login(email, password) {
        return this.uw
            .post('/auth/login', {
            email,
            password,
        })
            .then((response) => {
            this.onAuthenticated(response.meta.jwt, response.meta.socketToken);
            return helpers_1.parseDates(response.data, Auth.USER_DATE_FIELDS);
        });
    }
    logout() {
        return this.uw
            .delete('/auth')
            .then(() => null);
    }
}
exports.default = Auth;
Auth.USER_DATE_FIELDS = ['lastSeenAt', 'createdAt', 'updatedAt'];
